### PR TITLE
Implement `sf::Image::saveToMemory`

### PR DESCRIPTION
**Pull request by ChrisThrasher**
Implements https://github.com/SFML/SFML/pull/1669

This one got complicated since we have to take a `std::vector` and convert it into something that can be returned from a C function. I hope I did it well. It does incur the cost of copying the entire buffer which is unfortunate but there is no way to tell a vector to relinquish control of its contents so copying the vector into a `malloc`'d buffer seems like my only option.

### CODE REVIEW

File: src/SFML/Graphics/Image.cpp
Comment (eXpl0it3r) [inline src/SFML/Graphics/Image.cpp:145]: Wonder if we could/should assume the buffer to already be properly prepared. 🤔 

Like from a user perspective, I wonder whether one would really pass in some raw pointers and expect the function to allocate space for it.

Comment (ChrisThrasher) [reply src/SFML/Graphics/Image.cpp:145]: I believe the usage looks like this:
```c
unsigned char* buffer;
size_t size;
sfImage_saveToFile(image, &buffer, &size, "someformatstring");
```
Given C's limitations, this was the best API I could come up with. The problem with requiring users to prepare the buffer is that I'm not sure how they're supposed to correctly size that buffer. What I wrote means users don't have to know or care what the size it beforehand.

Comment (dogunbound) [reply src/SFML/Graphics/Image.cpp:145]: I think it's fine to malloc here. Requiring the user the do a malloc would in a sense, remove the "Simple" in SFML IMHO.

**Review by dogunbound — APPROVED** [review]
I really tried looking for a way to save stuff to the memory in sdl2. It doesn't seem like that have a way??? Maybe I didn't look hard enough. Anyways, documentation of buffer is all that's needed for me.

**Comment by kimci86** [comment]
> I really tried looking for a way to save stuff to the memory in sdl2. It doesn't seem like that have a way???

It is possible using a `SDL_RWops` object created with [SDL_RWFromMem](https://wiki.libsdl.org/SDL2/SDL_RWFromMem) and saving the image with [SDL_SaveBMP_RW](https://wiki.libsdl.org/SDL2/SDL_SaveBMP_RW). It requires to have a large enough memory buffer beforehand. I guess it is also possible to write to a dynamic buffer by using a `SDL_RWops`with custom callbacks.

**Comment by dogunbound** [comment]
> > I really tried looking for a way to save stuff to the memory in sdl2. It doesn't seem like that have a way???
> 
> It is possible using a [`SDL_RWops`](https://wiki.libsdl.org/SDL2/SDL_RWops) object created with [`SDL_RWFromMem`](https://wiki.libsdl.org/SDL2/SDL_RWFromMem) and saving the image with [`SDL_SaveBMP_RW`](https://wiki.libsdl.org/SDL2/SDL_SaveBMP_RW). It requires to have a large enough memory buffer beforehand. I guess it is also possible to write to a dynamic buffer by using a `SDL_RWops` with custom callbacks.

If that's true, I think my other suggestion where you have another function malloc the buffer before using it would be most similar.

**Comment by kimci86** [comment]
What about returning the vector in a C wrapper? This allows to avoid the copy of SFML's output into a new buffer. I explored the idea here: https://github.com/SFML/CSFML/compare/image_save_to_memory...kimci86:CSFML:image_save_to_memory

**Comment by eXpl0it3r** [comment]
What might also be an interesting thought process is, how this can be called from other languages, e.g. how it would be implemented in SFML.Net

**Comment by ChrisThrasher** [comment]
> What about returning the vector in a C wrapper? This allows to avoid the copy of SFML's output into a new buffer. I explored the idea here: [06faf7b](https://github.com/SFML/CSFML/commit/06faf7b51392991f051a4611a867b93d82b005e1)

I'm open to the idea, but I'm not in love with it since it adds quite a bit of code that we currently would only use in one place. Is there a 2nd location in CSFML or any speculated future uses of this abstraction?

File: src/SFML/Graphics/Image.cpp
Comment (oprypin) [reply src/SFML/Graphics/Image.cpp:145]: It's impossible for the user to predict how big the buffer should be. After all, image compression makes it so the data size can be anything whatsoever. I think this solution as suggested is the best that can be done with the given API.

Comment (oprypin) [reply src/SFML/Graphics/Image.cpp:145]: But it is true that this is the first ever usage of `malloc` in this repository (previously it was all `new`). And it's possible that the user is using yet another kind of allocator but will need to deal with `free` specifically only for this use case.

**Comment by oprypin** [comment]
> What about returning the vector in a C wrapper? This allows to avoid the copy of SFML's output into a new buffer. I explored the idea here: [06faf7b](https://github.com/SFML/CSFML/commit/06faf7b51392991f051a4611a867b93d82b005e1)

This is actually a great idea, *but* it shouldn't be *returned*, it should still be passed as an output arg and instantiated by the user, then we'd also eliminate *all* copying and leave all mallocs up to the user. There's also an issue with your commit that the `new` allocator is used but the user then can't correctly `delete` the memory from C land.

As such, my suggestion would be a slight edit to your commit, then I like how it looks:
https://github.com/oprypin/CSFML/commit/e4f82f0df9080c87a347f6bcee6bdc6bcb0e3fdf

Or how it looks all together:
https://github.com/oprypin/CSFML/compare/a2c73376f96dca6ecd2034aeac4e8a4e22087830~3...e4f82f0df9080c87a347f6bcee6bdc6bcb0e3fdf
(sure would be neat if this branch was now merged as a PR authored by 3 people :innocent:)

I think with this, all the mentioned concerns about allocation are eliminated.

Maybe this new type should be moved out of Graphics module, though?

**Comment by oprypin** [comment]
I am not directly using CSFML at all, but I am saying that the approach I took there is like this proposed alternative:

> Or how it looks all together:
> [oprypin/csfml@e4f82f0~3...d8281df](https://github.com/oprypin/csfml/compare/e4f82f0df9080c87a347f6bcee6bdc6bcb0e3fdf~3...d8281dfdb5c168dd291b189919ec945098145521)

**Comment by ChrisThrasher** [comment]
This is one of the least CSFML 2.6.0 APIs to implement before we can release it. It's important we not stall the release for too much longer. Can we accept the PR in its current state since honestly this interface is unlikely to get used very much to begin with?

If we want to relitigate these decisions while developing CSFML 3 then so be it but I'd rather not let this PR sit for more weeks or months further stalling shipping CSFML 2.6.0.

**Comment by oprypin** [comment]
We have named several disadvantages of this PR in the current state, which https://github.com/oprypin/csfml/compare/e4f82f0df9080c87a347f6bcee6bdc6bcb0e3fdf~3...d8281dfdb5c168dd291b189919ec945098145521 doesn't have. Why not merge it and discuss from there? Maybe that's even the reason that the PR is stalled - a solution was proposed but not acknowledged

**Comment by ChrisThrasher** [comment]
I don't think we need something as complex as what you proposed. The reality is that we're letting a relatively unimportant PR stall an entire release. This API is not important enough to warrant this much attention. If someone wants to submit an alternative PR (with all the right docs already written and passing CI) then we can consider an alternative but so far nobody has stepped up to do that so this is our only option at the moment.

**Comment by oprypin** [comment]
This method would lose most of its useful properties this way. If it's so unimportant then why was it added to SFML in the first place?

Again, let me list the deficiencies:

* The user must guess the buffer size, or probably provide a buffer that is surely too big for the destination image, or else call this method twice to get the correct size first.
* One cannot reuse an existing buffer, it must be allocated on every call, on the heap and then disposed manually.
* It is the first ever use of `malloc` in this library (adding another allocator just for this) **and the first time the user is forced to de-allocate memory specifically using C standard library.**
* Bindings to most other languages will be forced to re-copy this data in order to expose it back to the target language.

You mention "complexity" and "abstraction", but until C has anonymous tuples or something, putting data into a struct of 2 members is as simple as it gets. Plus, that struct directly mirrors the single `vector` parameter just like it is in the C++ API.

**Comment by ChrisThrasher** [comment]
> The user must guess the buffer size

This is not true. The API does not require preallocating anything or having to guess any sizes beforehand.

Are you willing to write up a PR for your proposed implementation?

**Comment by oprypin** [comment]
Sorry, you're right, it does not require guessing

**Review by dogunbound — APPROVED** [review]
As long as we at least document what is getting freed and what the user has to manage, then LGTM!!!

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/CSFML/blob/a80542d67d86dd37b9c6693e11ee666cbcfbe9f1/src/SFML/Window/VideoMode.cpp#L47-L69

Perhaps we copy what something like `sfVideoMode_getFullscreenModes` is doing. It internally creates a static `std::vector` and then assigns into that vector before finally returning the address of the first member of that vector and the size is an out parameter. This means so long as you don't call `sfVideoMode_getFullscreenModes` a 2nd time you can keep using the pointer that it returned.

**Comment by ChrisThrasher** [comment]
```cpp
sfBool sfImage_saveToMemory(const sfImage* image, unsigned char** buffer, size_t* size, const char* format)
{
    CSFML_CHECK_RETURN(image, sfFalse);

    static std::vector<sf::Uint8> output;
    if (image->This.saveToMemory(output, format)){
        // Create new buffer then return address and size to user
        *buffer = &output[0];
        *size = output.size();
        return sfTrue;
    }

    *buffer = NULL;
    *size = 0;
    return sfFalse;
}
```
This removes the `malloc` and better matches how other CSFML APIs behave.

**Comment by oprypin** [comment]
Ah no, that would just mean that you can't keep two images at once and look at them

**Comment by ChrisThrasher** [comment]
I might be coming around to having an `sfBuffer` struct. Are you willing to write docs for all the new APIs you added? If so then we can merge that into this PR and discuss from there.

**Comment by oprypin** [comment]
I looked into it briefly, but sorry, I'm not willing to write that boilerplate. Actually there are a lot of small but still non-trivial decisions in creating the new files for this new class

**Comment by kimci86** [comment]
I would be happy to write the clean implementation with sfBuffer. I expect to have time do it by the end of the week.

**Comment by ChrisThrasher** [comment]
I'd love to see that 👍🏻

**Comment by kimci86** [comment]
Here we go: https://github.com/SFML/CSFML/compare/image_save_to_memory...kimci86:CSFML:image_save_to_memory

**Comment by ChrisThrasher** [comment]
That's a clean solution now that I really grasp what's going on. Can't argue against that. If you open a PR I'll close this one.

**Comment by ChrisThrasher** [comment]
#205 is better. Let's focus on that PR instead.

**Pull request closed without merging**
